File: stage2/mc_rainbow.cpp
// multithreaded Monte Carlo pricer for rainbow option: (max(S1,...,Sn) - K)+
// correlated GBM via Cholesky, OpenMP parallelism
// compile: g++ -O3 -std=c++17 -fopenmp mc_rainbow.cpp -o mc_rainbow -lm

#include <iostream>
#include <vector>
#include <cmath>
#include <random>
#include <numeric>
#include <algorithm>
#include <chrono>
#include <thread>
#include <iomanip>

using Vec = std::vector<double>;
using Mat = std::vector<Vec>;

Mat cholesky(const Mat& A) { // lower triangular L such that A = L*L^T
    int n = A.size();
    Mat L(n, Vec(n, 0.0));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j <= i; j++) {
            double sum = 0.0;
            for (int k = 0; k < j; k++) sum += L[i][k] * L[j][k];
            if (i == j) L[i][j] = std::sqrt(A[i][i] - sum);
            else L[i][j] = (A[i][j] - sum) / L[j][j];
        }
    }
    return L;
}

struct MCResult {
    double price;
    double stderr;
    double elapsed_ms;
};

MCResult mc_rainbow(const Vec& S0, double K, double T, double r,
                    const Vec& sigmas, const Mat& corr, int n_paths, int n_threads) {
    int d = S0.size();
    Mat L = cholesky(corr);

    // precompute drifts
    Vec drift(d);
    for (int i = 0; i < d; i++)
        drift[i] = (r - 0.5 * sigmas[i] * sigmas[i]) * T;

    double sqrt_T = std::sqrt(T);
    double discount = std::exp(-r * T);

    // per-thread accumulators
    std::vector<double> thread_sum(n_threads, 0.0);
    std::vector<double> thread_sum2(n_threads, 0.0);
    int paths_per_thread = n_paths / n_threads;

    auto start = std::chrono::high_resolution_clock::now();

    std::vector<std::thread> threads;
    for (int t = 0; t < n_threads; t++) {
        threads.emplace_back([&, t]() {
            std::mt19937_64 rng(42 + t); // different seed per thread
            std::normal_distribution<double> norm(0.0, 1.0);

            Vec Z(d), W(d), S_T(d);
            double local_sum = 0.0, local_sum2 = 0.0;

            for (int p = 0; p < paths_per_thread; p++) {
                // generate independent normals
                for (int i = 0; i < d; i++) Z[i] = norm(rng);

                // correlate via Cholesky: W = L * Z
                for (int i = 0; i < d; i++) {
                    W[i] = 0.0;
                    for (int j = 0; j <= i; j++) W[i] += L[i][j] * Z[j];
                }

                // terminal stock prices
                double max_S = 0.0;
                for (int i = 0; i < d; i++) {
                    S_T[i] = S0[i] * std::exp(drift[i] + sigmas[i] * sqrt_T * W[i]);
                    max_S = std::max(max_S, S_T[i]);
                }

                double payoff = std::max(max_S - K, 0.0);
                local_sum += payoff;
                local_sum2 += payoff * payoff;
            }

            thread_sum[t] = local_sum;
            thread_sum2[t] = local_sum2;
        });
    }

    for (auto& th : threads) th.join();

    auto end = std::chrono::high_resolution_clock::now();
    double elapsed = std::chrono::duration<double, std::milli>(end - start).count();

    int total_paths = paths_per_thread * n_threads;
    double sum = 0.0, sum2 = 0.0;
    for (int t = 0; t < n_threads; t++) {
        sum += thread_sum[t];
        sum2 += thread_sum2[t];
    }

    double mean_payoff = sum / total_paths;
    double var = sum2 / total_paths - mean_payoff * mean_payoff;
    double price = discount * mean_payoff;
    double stderr = discount * std::sqrt(var / total_paths);

    return {price, stderr, elapsed};
}

int main() {
    int n_threads = std::thread::hardware_concurrency();
    std::cout << "Threads: " << n_threads << "\n\n";

    // --- 2-asset ---
    {
        Vec S0 = {100, 100};
        Vec sigmas = {0.2, 0.3};
        Mat corr = {{1.0, 0.5}, {0.5, 1.0}};

        std::cout << "=== 2-ASSET RAINBOW (10M paths) ===\n";
        auto r = mc_rainbow(S0, 100, 1.0, 0.05, sigmas, corr, 10'000'000, n_threads);
        std::cout << std::fixed << std::setprecision(4);
        std::cout << "Price:  " << r.price << " ± " << r.stderr << "\n";
        std::cout << "Time:   " << std::setprecision(1) << r.elapsed_ms << " ms\n\n";
    }

    // --- 5-asset ---
    {
        Vec S0(5, 100);
        Vec sigmas = {0.2, 0.25, 0.3, 0.22, 0.28};
        int d = 5;
        Mat corr(d, Vec(d, 0.3));
        for (int i = 0; i < d; i++) corr[i][i] = 1.0;

        std::cout << "=== 5-ASSET RAINBOW (10M paths) ===\n";
        auto r = mc_rainbow(S0, 100, 1.0, 0.05, sigmas, corr, 10'000'000, n_threads);
        std::cout << std::fixed << std::setprecision(4);
        std::cout << "Price:  " << r.price << " ± " << r.stderr << "\n";
        std::cout << "Time:   " << std::setprecision(1) << r.elapsed_ms << " ms\n\n";
    }

    // --- 20-asset ---
    {
        int d = 20;
        Vec S0(d, 100);
        Vec sigmas(d);
        for (int i = 0; i < d; i++) sigmas[i] = 0.15 + 0.01 * i; // 0.15 to 0.34
        Mat corr(d, Vec(d, 0.3));
        for (int i = 0; i < d; i++) corr[i][i] = 1.0;

        std::cout << "=== 20-ASSET RAINBOW (10M paths) ===\n";
        auto r = mc_rainbow(S0, 100, 1.0, 0.05, sigmas, corr, 10'000'000, n_threads);
        std::cout << std::fixed << std::setprecision(4);
        std::cout << "Price:  " << r.price << " ± " << r.stderr << "\n";
        std::cout << "Time:   " << std::setprecision(1) << r.elapsed_ms << " ms\n\n";
    }

    // --- 50-asset ---
    {
        int d = 50;
        Vec S0(d, 100);
        Vec sigmas(d);
        for (int i = 0; i < d; i++) sigmas[i] = 0.15 + 0.005 * i;
        Mat corr(d, Vec(d, 0.3));
        for (int i = 0; i < d; i++) corr[i][i] = 1.0;

        std::cout << "=== 50-ASSET RAINBOW (10M paths) ===\n";
        auto r = mc_rainbow(S0, 100, 1.0, 0.05, sigmas, corr, 10'000'000, n_threads);
        std::cout << std::fixed << std::setprecision(4);
        std::cout << "Price:  " << r.price << " ± " << r.stderr << "\n";
        std::cout << "Time:   " << std::setprecision(1) << r.elapsed_ms << " ms\n\n";
    }

    // --- 100-asset ---
    {
        int d = 100;
        Vec S0(d, 100);
        Vec sigmas(d);
        for (int i = 0; i < d; i++) sigmas[i] = 0.15 + 0.002 * i; // 0.15 to 0.348
        Mat corr(d, Vec(d, 0.3));
        for (int i = 0; i < d; i++) corr[i][i] = 1.0;

        std::cout << "=== 100-ASSET RAINBOW (10M paths) ===\n";
        auto r = mc_rainbow(S0, 100, 1.0, 0.05, sigmas, corr, 10'000'000, n_threads);
        std::cout << std::fixed << std::setprecision(4);
        std::cout << "Price:  " << r.price << " ± " << r.stderr << "\n";
        std::cout << "Time:   " << std::setprecision(1) << r.elapsed_ms << " ms\n";
    }

    return 0;
}

File: stage2/mc_rainbow.py
# Monte Carlo pricer for multi-asset rainbow option
# payoff: (max(S1, ..., Sn) - K)+
# uses correlated GBM with Cholesky decomposition
# works for arbitrary number of assets

import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import time


def mc_rainbow(S0, K, T, r, sigmas, corr_matrix, n_paths=1_000_000, seed=42):
    rng = np.random.default_rng(seed)
    n_assets = len(S0)
    L = np.linalg.cholesky(corr_matrix)  # correlate the brownian motions

    Z = rng.standard_normal((n_paths, n_assets))
    W = Z @ L.T  # correlated normals

    # GBM terminal values: S_T = S0 * exp((r - 0.5*sigma^2)*T + sigma*sqrt(T)*W)
    sigmas = np.array(sigmas)
    S0 = np.array(S0)
    drift = (r - 0.5 * sigmas**2) * T
    diffusion = sigmas * np.sqrt(T) * W
    S_T = S0 * np.exp(drift + diffusion)  # (n_paths, n_assets)

    payoff = np.maximum(np.max(S_T, axis=1) - K, 0)  # rainbow payoff
    price = np.exp(-r * T) * np.mean(payoff)
    stderr = np.exp(-r * T) * np.std(payoff) / np.sqrt(n_paths)

    return price, stderr, S_T, payoff


def mc_rainbow_greeks(S0, K, T, r, sigmas, corr_matrix, n_paths=1_000_000, bump=1.0):
    # finite difference greeks via bump-and-reprice
    n_assets = len(S0)
    price_base, _, _, _ = mc_rainbow(S0, K, T, r, sigmas, corr_matrix, n_paths)

    deltas = np.zeros(n_assets)
    gammas = np.zeros(n_assets)

    for i in range(n_assets):
        S_up = S0.copy(); S_up[i] += bump
        S_dn = S0.copy(); S_dn[i] -= bump

        p_up, _, _, _ = mc_rainbow(S_up, K, T, r, sigmas, corr_matrix, n_paths)
        p_dn, _, _, _ = mc_rainbow(S_dn, K, T, r, sigmas, corr_matrix, n_paths)

        deltas[i] = (p_up - p_dn) / (2 * bump)
        gammas[i] = (p_up - 2 * price_base + p_dn) / (bump**2)

    return price_base, deltas, gammas


def build_corr_matrix(n_assets, rho=0.3):  # uniform pairwise correlation
    corr = np.full((n_assets, n_assets), rho)
    np.fill_diagonal(corr, 1.0)
    return corr


if __name__ == '__main__':
    # 2-asset case — compare against FDM
    print('=' * 50)
    print('2-ASSET RAINBOW — MC vs FDM comparison')
    print('=' * 50)
    S0_2 = np.array([100.0, 100.0])
    sigmas_2 = [0.2, 0.3]
    corr_2 = np.array([[1.0, 0.5], [0.5, 1.0]])

    t0 = time.time()
    price_2, stderr_2, _, _ = mc_rainbow(S0_2, 100, 1.0, 0.05, sigmas_2, corr_2, n_paths=10_000_000)
    t_2 = time.time() - t0
    print(f'MC price (10M paths): {price_2:.4f} ± {stderr_2:.4f}')
    print(f'FDM price:            18.7773 (from bs_fdm_2asset.py)')
    print(f'Difference:           {abs(price_2 - 18.7773):.4f}')
    print(f'Wall clock:           {t_2:.2f}s')

    # greeks for 2-asset
    print('\nGreeks (2-asset):')
    _, deltas_2, gammas_2 = mc_rainbow_greeks(S0_2, 100, 1.0, 0.05, sigmas_2, corr_2, n_paths=2_000_000)
    for i in range(2):
        print(f'  Asset {i+1}: Delta = {deltas_2[i]:.4f}, Gamma = {gammas_2[i]:.6f}')

    # 5-asset case — the actual target
    print('\n' + '=' * 50)
    print('5-ASSET RAINBOW — Ground Truth')
    print('=' * 50)
    S0_5 = np.array([100.0, 100.0, 100.0, 100.0, 100.0])
    sigmas_5 = [0.2, 0.25, 0.3, 0.22, 0.28]
    corr_5 = build_corr_matrix(5, rho=0.3)

    t0 = time.time()
    price_5, stderr_5, S_T_5, payoff_5 = mc_rainbow(S0_5, 100, 1.0, 0.05, sigmas_5, corr_5, n_paths=10_000_000)
    t_5 = time.time() - t0
    print(f'MC price (10M paths): {price_5:.4f} ± {stderr_5:.4f}')
    print(f'Wall clock:           {t_5:.2f}s')

    # greeks for 5-asset
    print('\nGreeks (5-asset):')
    _, deltas_5, gammas_5 = mc_rainbow_greeks(S0_5, 100, 1.0, 0.05, sigmas_5, corr_5, n_paths=2_000_000)
    for i in range(5):
        print(f'  Asset {i+1} (σ={sigmas_5[i]}): Delta = {deltas_5[i]:.4f}, Gamma = {gammas_5[i]:.6f}')

    # ITM/OTM analysis
    itm_mask = payoff_5 > 0
    print(f'\nITM probability: {np.mean(itm_mask):.4f}')
    print(f'Mean payoff (ITM only): {np.mean(payoff_5[itm_mask]):.4f}')

    # plot payoff distribution
    fig, axes = plt.subplots(1, 2, figsize=(14, 5))

    axes[0].hist(payoff_5[payoff_5 > 0], bins=100, density=True, alpha=0.7, color='steelblue')
    axes[0].set_xlabel('Payoff')
    axes[0].set_ylabel('Density')
    axes[0].set_title('5-Asset Rainbow: Payoff Distribution (ITM only)')
    axes[0].axvline(price_5 * np.exp(0.05), color='red', linestyle='--', label=f'Undiscounted price')
    axes[0].legend()

    # convergence plot
    path_counts = [1000, 5000, 10000, 50000, 100000, 500000, 1000000, 5000000, 10000000]
    prices_conv = []
    for n in path_counts:
        p, _, _, _ = mc_rainbow(S0_5, 100, 1.0, 0.05, sigmas_5, corr_5, n_paths=n)
        prices_conv.append(p)

    axes[1].semilogx(path_counts, prices_conv, 'o-', linewidth=2)
    axes[1].axhline(price_5, color='red', linestyle='--', alpha=0.5, label=f'10M estimate: {price_5:.4f}')
    axes[1].set_xlabel('Number of Paths')
    axes[1].set_ylabel('Price')
    axes[1].set_title('MC Convergence (5-Asset Rainbow)')
    axes[1].legend()
    axes[1].grid(True, alpha=0.3)

    plt.tight_layout()
    plt.savefig('mc_rainbow_analysis.png', dpi=150)
    print('\nSaved: mc_rainbow_analysis.png')
